DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,269 @@
+
+### Writing Test Cases
+
+When you implement a feature, unit tests are required.
+
+Most Harvester features follow the [Kubernetes controller pattern](https://kubernetes.io/docs/concepts/architecture/controller/), so it's important to test them. Please take a look at the following example to understand basic usage:
+- [support bundle controller](./pkg/controller/master/supportbundle/controller_test.go)
+
+In addition to unit tests, write automated tests where applicable. If the feature requires an automated test, refer to the [Automation Testing](#automation-testing) section for how to write and run them.
+
+## Debugging and Troubleshooting
+
+In general, we need cluster information for debugging. We usually obtain a [support bundle](https://docs.harvesterhci.io/v1.6/troubleshooting/harvester/#generate-a-support-bundle) from users. It contains a lot of useful cluster information.
+
+Then, we use [rancher/support-bundle-kit](https://github.com/rancher/support-bundle-kit) to analyze the support bundle. It's a toolkit for generating and analyzing support bundles for Kubernetes and Kubernetes-native applications.
+
+After building the binary from the project, you can run `./bin/support-bundle-kit-{amd64|arm64} simulator --bundle-path ./supportbundle_xxxx`. It takes some time to build the simulated cluster. Then you can use the kubeconfig from `~/.sim/admin.kubeconfig` to operate the cluster. Since it reconstructs the cluster only from YAML files and logs, commands inside pods will not work.
+
+Finally, refer to the following resources to search for possible solutions.
+- [Harvester Documentation](https://docs.harvesterhci.io/v1.6/)
+- [GitHub Issues](https://github.com/harvester/harvester/issues)
+- [Harvester HCI Knowledge Base](https://harvesterhci.io/kb/)
+
+## Testing and Building
+
+All repositories use Dapper to build. Ensure you have a Docker-compatible container engine before building.
+
+### Test Your Changes by Patching the Image
+
+For quick testing, you can patch the image in the cluster with a locally built image. Normally, you can use `make` or `./scripts/build` in each repository, and push the result to a registry by retagging the resulting image. If that doesn't work, see the scripts under `./scripts`.
+
+In general, each repository includes two components:
+
+- Controller
+- Webhook
+
+Therefore, if you develop a virtual machine feature, you should patch the image in the Harvester controller. This depends on which feature you're developing. Before patching, identify which repositories need to be updated in the Harvester cluster.
+
+Please check [Repository Relations](#repository-relations) for more details. If you're not sure, leave a comment in your issue.
+
+> [!IMPORTANT]  
+> Not all features/repositories apply to this section, such as harvester/harvester-installer.  
+> For those, use the next section to build a fresh ISO for testing.  
```

Comment:
I don't think these 2 paragraphs are necessary. If we want to point out that developing the installer requires a different workflow to produce the iso, then just transition straight to the next section, after explaining the controller and webhook images.